### PR TITLE
Properly handle subtree write rules (do not allow non empty subtree write rules)

### PR DESCRIPTION
**Pull request by platfowner**
Change summary:
- Properly handle subtree write rules (do not allow non empty subtree write rules)
- Return more informative error_message for false rule eval
- Fix a bug on subtree rules
- Tweak checkout history rule location (moved under 'data')
- Tweak sharding setup tx operation order
- Use setValuesForTesting() instead of writeDatabase() or setValue() in tests
- Add missing result code
- Reorg db unit tests
- Exclude error_message from receipts
- Add rebuild_genesis_blocks.sh tool script

Related issues:
- https://github.com/ainblockchain/ain-blockchain/issues/779

### CODE REVIEW

File: db/index.js [hunk 1]
```diff
@@ -2073,38 +2103,63 @@ class DB {
...
+        evalResult: false,
+      };
     }
-    const writeRuleString = writeRuleConfig[RuleProperties.WRITE];
-    if (typeof writeRuleString === 'boolean') {
-      return writeRuleString;
-    } else if (typeof writeRuleString !== 'string') {
-      return false;
+    const ruleString = writeRuleConfig[RuleProperties.WRITE];
+    if (typeof ruleString === 'boolean') {
```
Comment (liayoo) [inline db/index.js:2112]: CommonUtil.isBool()?

File: db/index.js [hunk 2]
```diff
@@ -2073,38 +2103,63 @@ class DB {
...
+      return {
+        ruleString: String(ruleString),
+        evalResult: ruleString,
+      };
+    } else if (typeof ruleString !== 'string') {
```
Comment (liayoo) [inline db/index.js:2117]: CommonUtil.isString()?

File: db/index.js [hunk 3]
```diff
@@ -1719,30 +1715,63 @@ class DB {
...
-      if (!evalWriteRuleRes) {
-        logger.debug(`[${LOG_HEADER}] evalWriteRuleRes ${evalWriteRuleRes}, ` +
+      if (!evalWriteRuleRes.evalResult) {
+        logger.debug(`[${LOG_HEADER}] evalWriteRuleRes ${JSON.stringify(evalWriteRuleRes, null, 2)}, ` +
             `matched: ${JSON.stringify(matchedWriteRules, null, 2)}, data: ${JSON.stringify(data)}, ` +
             `newData: ${JSON.stringify(newData)}, auth: ${JSON.stringify(auth)}, ` +
             `timestamp: ${timestamp}\n`);
+        return {
+          code: TxResultCode.VALUE_PERMISSION_FALSE_WRITE_RULE_EVAL,
+          error_message: `False eval of write rule [${evalWriteRuleRes.ruleString}] ` +
```
[the remark this answers is not in the document]
Comment (platfowner) [reply db/index.js]: As discussed offline, just removed error messages from the receipts as now we're using informative code.
And tried to tweak the prefix of the error messages. PTAL

File: db/index.js [hunk 4]
```diff
@@ -1719,30 +1715,63 @@ class DB {
...
-      evalStateRuleRes = this.evalStateRuleConfig(matchedStateRules.closestRule.config, newValue);
-      if (!evalStateRuleRes) {
+      const evalStateRuleRes = this.evalStateRuleConfig(matchedStateRules.closestRule.config, newValue);
+      if (!evalStateRuleRes.evalResult) {
         logger.debug(`[${LOG_HEADER}] evalStateRuleRes ${evalStateRuleRes}, ` +
             `matched: ${JSON.stringify(matchedStateRules, null, 2)}, parsedValuePath: ${parsedValuePath}, ` +
             `newValue: ${JSON.stringify(newValue)}\n`);
+        return {
+          code: TxResultCode.VALUE_PERMISSION_FALSE_STATE_RULE_EVAL,
+          error_message: `False eval of state rule [${evalStateRuleRes.ruleString}] ` +
```
[the remark this answers is not in the document]
Comment (platfowner) [reply db/index.js]: ditto~

File: db/index.js [hunk 5]
```diff
@@ -1719,30 +1715,63 @@ class DB {
         this.addPathToValue(value, matchedWriteRules.matchedValuePath, matchedWriteRules.closestRule.path.length);
```
[the remark this answers is not in the document]
Comment (platfowner) [reply db/index.js:1712]: Just removed, thanks!

File: rebuild_genesis_blocks.sh [hunk 6]
```diff
@@ -0,0 +1,66 @@
+#!/bin/bash
+
+printf "\n[[[[[ rebuild_genesis_blocks.sh ]]]]]\n\n"
```
[the remark this answers is not in the document]
Comment (platfowner) [reply rebuild_genesis_blocks.sh:3]: Thanks, this is just a quick solution. Feel free to update or replace with better ones.

File: db/index.js (hunk 1 above)
Comment (platfowner) [reply db/index.js]: Done, thanks!

File: db/index.js (hunk 2 above)
Comment (platfowner) [reply db/index.js]: Done, thanks!

**Comment by platfowner** [comment]
Comments address, PTAL~ @liayoo

File: p2p/server.js [hunk 7]
```diff
@@ -868,31 +868,38 @@ class P2pServer {
         type: WriteDbOperations.SET,
         op_list: [
           {
-            type: WriteDbOperations.SET_RULE,
```
Comment (liayoo) [inline p2p/server.js:871]: question) I noticed that the order of txs is changed. Just curious, but does the order affect the results of the transactions?

File: test/unit/db.test.js [hunk 8]
```diff
@@ -137,8 +135,7 @@ describe("DB operations", () => {
         }
       }
     };
-    result = node.db.setValue("/apps/test", dbValues);
-    assert.deepEqual(result.code, 0);
+    node.db.setValuesForTesting("/apps/test", dbValues);
```
Comment (liayoo) [inline test/unit/db.test.js]: 👍

File: db/index.js [hunk 9]
```diff
@@ -1719,30 +1712,63 @@ class DB {
...
     const newData =
         this.addPathToValue(newValue, matchedWriteRules.matchedValuePath, matchedWriteRules.closestRule.path.length);
-    let evalWriteRuleRes = false;
-    let evalStateRuleRes = false;
+    if (matchedWriteRules.subtreeRules && matchedWriteRules.subtreeRules.length > 0) {
+      return {
+        code: TxResultCode.VALUE_PERMISSION_NON_EMPTY_SUBTREE_RULES,
+        error_message: `Non-empty (${matchedWriteRules.subtreeRules.length}) subtree rules on: ` +
```
Comment (liayoo) [inline db/index.js:1718]: minor) on -> at?

**Review by liayoo — APPROVED** [review]
LGTM with some minor comments & a question. Thanks!

File: p2p/server.js (hunk 7 above)
Comment (platfowner) [reply p2p/server.js:871]: right, the SET_RULE operations creates subtree rules so they're moved after the SET_VALUE operation.
just added some notes for readers.

File: db/index.js (hunk 9 above)
Comment (platfowner) [reply db/index.js]: just replaced by 'for value path', thanks!

**Comment by platfowner** [comment]
Thanks for the review, now merging...

File: p2p/server.js (hunk 7 above)
Comment (liayoo) [reply p2p/server.js:871]: I see! Thanks for the explanation & the note.

**Pull request merged**